### DISCUSSION TITLE
Error Trial level node is missing the "type" parameter. The parameters for the node are: {}

### DISCUSSION THREAD

**Discussion by dcousin3** [jspsych/jsPsych#931] [Q&A] [answered]
I put two elements in a timeline of the experience and things work great. However, when I make a separate timeline with the same two elements, I get the error :
`Trial level node is missing the "type" parameter. The parameters for the node are: {}`

Here is a (near) minimum working example:
```
<!DOCTYPE html>
<html>
    <head>
        <!-- character encoding -->
        <meta content="text/html;charset=utf-8" http-equiv="Content-Type">
        <meta content="utf-8" http-equiv="encoding">
        <!-- Title of the window -->
        <title>MWE</title>
        <!-- declaring the special functions needed -->
        <!-- note that jspsych-html-keyboard-response-noerase.js IS NOT part of the standard bundle -->
        <script src="jspsych-6.1.0/jspsych.js"></script>
        <script src="jspsych-6.1.0/plugins/jspsych-html-keyboard-response.js"></script>
        <script src="jspsych-6.1.0/plugins/jspsych-audio-button-response.js"></script>
        <script src="jspsych-6.1.0/plugins/jspsych-html-button-response.js"></script>
        <script src="jspsych-6.1.0/plugins/jspsych-call-function.js"></script>
        <!-- uploading the style sheet -->
        <link href="jspsych-6.1.0/css/jspsych.css" rel="stylesheet" type="text/css"></link>
    </head>


    <body>
        <script>
            <!-- ****************************** -->
            <!-- *   Defining general stuff   * -->
            <!-- ****************************** -->
            
            var OneItemHTML       = '<svg width="100" height="100"><rect width="100" height="100" style="fill:rgb(0,0,0);stroke-width:0;" /></svg>';
            var OneItemACKNW      = '<svg width="100" height="100"><rect width="100" height="100" style="fill:rgb(126,126,126);stroke-width:0;" /></svg>';
            var RecallSignal       = 'https://raw.githubusercontent.com/RMG2424/Captain-Marvel/master/500hz-400ms.wav'

            <!-- Empty lists of items; are to be generated by myfunc -->
            var ListOfItems         = [];
            var ListOfBlinkingItems = [];
            var ListOfButtons       = [];

            ListOfItems.push( `<p style = "position:absolute; margin-top: 0em; left: 100px; top: 100px">${OneItemACKNW}</p>` );
            ListOfItems.push( `<p style = "position:absolute; margin-top: 0em; left: 200px; top: 200px">${OneItemACKNW}</p>` );
            
            ListOfButtons.push(`<button type=button style = "position:absolute; margin: 0px 0px; padding: 0px 0px; border: 0px; left: 100px; top: 100px">${OneItemHTML}</button>` )
            ListOfButtons.push(`<button type=button style = "position:absolute; margin: 0px 0px; padding: 0px 0px; border: 0px; left: 200px; top: 200px">${OneItemHTML}</button>` )

            var ReadAllResponses = {
                timeline: [ ReadOneResponse, BlinkThatResponse ]
            }

            var ReadOneResponse = {
                type: 'audio-button-response',
                stimulus: RecallSignal,
                button_html: ListOfButtons,
                choices: function() {return [...Array(ListOfButtons.length).keys()]},
                trial_duration: 5000,
                on_finish: function() {console.log(jsPsych.data.get().last(1).values()[0].button_pressed)}                
            }
            var BlinkThatResponse = {
                type: 'html-button-response',
                stimulus: '',
                button_html: function() {return ListOfButtons.concat(ListOfItems[jsPsych.data.get().last(1).values()[0].button_pressed])},
                trial_duration: 1000,
                choices: function() {return [...Array(ListOfButtons.length+1).keys()]}
            }
            
            jsPsych.init({
<!-- this works                timeline: [ ReadOneResponse, BlinkThatResponse ] -->
                timeline: [ ReadAllResponses ] <!-- that does not -->
            })

        </script>
    </body>
</html>
```
Here I could add the two elements, but ultimately, I would like to have a timeline_variables to ReadAllResponses so that a variable number of responses can be read, e.g.
```
var ReadAllResponses = {
                timeline: [ ReadOneResponse, BlinkThatResponse ],
                timeline_variables: <<<a list of variable length>>>
          }
```

I must be doing something wrong, but can't figure what it is... Any help appreciated.

Many thanks!

**Comment by becky-gilbert** [comment]
**EDIT 2**: I think I this response is wrong, and I don't think there's a bug related to objects on the timeline that only contain their own timeline. See the selected answer. 
<hr>

**EDIT 1**: Apparently this is a bug rather than the intended behavior (see Josh's comment below and issue #938). When this bug is fixed, it should be possible to have an object that just contains its own timeline. But until then, this response might be helpful for others having the same problem.
<hr>

Hi @dcousin3! I think the problem here is that jsPsych doesn't know what to do with an object that only contains a timeline. The object containing the timeline also has to have some timeline variables to repeat over, or has a loop function or conditional function. Otherwise, if you want to just run the two trials, then you can put them inside of an array and use that as your main experiment timeline (as you have done in the commented-out line in your jsPsych.init):

```javascript
var ReadAllResponses = [ ReadOneResponse, BlinkThatResponse ];
```

Whereas if you want to put these trials into a procedure using timeline _variables, then you will need to include a timeline_variables array inside the object that contains your timeline of trials. It might work to include a timeline_variables array with an empty object, and without using `jsPsych.timelineVariable()` anywhere in your trials, but I haven't tried this!
```javascript
var ReadAllResponses = {
    timeline: [ ReadOneResponse, BlinkThatResponse ],
    timeline_variables: [{}]
}
```

**Reply by dcousin3** [reply]
Hello, 
Well, I must be damned, because I did that many times, to no avail, and now it works... Many thanks and sorry about the hassle.
Denis

**Reply by jodeleeuw** [reply]
This isn't intended behavior. It was supposed to be the case that just specifying a timeline would work. I'll have to add some tests to the library to see what the source of the issue is.

**Reply by becky-gilbert** [reply]
Ah, good to know @jodeleeuw. I've updated my response.

**Comment by becky-gilbert** [comment, marked as answer]
Coming back to this question because I just looked into this issue and I think I was wrong - I don't think there's a bug. You should be able to use an object in the timeline that just contains a timeline array and no other parameters. 

I think the reason that your code wasn't working before is because the `ReadAllResponses` object needs to be created _after_ the two trials that it contains have been defined. So it should work if you just reorder the definitions of the trials and the `ReadAllResponses` object:
```javascript
var ReadOneResponse = {
  type: 'audio-button-response',
  stimulus: RecallSignal,
  button_html: ListOfButtons,
  choices: function() {return [...Array(ListOfButtons.length).keys()]},
  trial_duration: 5000,
  on_finish: function() {console.log(jsPsych.data.get().last(1).values()[0].button_pressed)}                
}
var BlinkThatResponse = {
  type: 'html-button-response',
  stimulus: '',
  button_html: function() {return ListOfButtons.concat(ListOfItems[jsPsych.data.get().last(1).values()[0].button_pressed])},
  trial_duration: 1000,
  choices: function() {return [...Array(ListOfButtons.length+1).keys()]}
}
var ReadAllResponses = {
  timeline: [ ReadOneResponse, BlinkThatResponse ]
}
jsPsych.init({
  timeline: [ ReadAllResponses ] 
})
```
So I'm going to unmark my other response and mark this one as the answer. But please let me know if you get a chance to try this and it doesn't solve the problem!

**Comment by dcousin3** [comment]
excellent,

Thanks for looking into it.

denis.

De : Becky Gilbert [mailto:notifications@github.com]
Envoyé : 18 octobre 2020 17:39
À : jspsych/jsPsych
Cc : Denis Cousineau; Mention
Objet : Re: [jspsych/jsPsych] Error Trial level node is missing the "type" parameter. The parameters for the node are: {} (#931)

Attention : courriel externe | external email

Coming back to this question because I just looked into this issue and I think I was wrong - I don't think there's a bug. You should be able to use an object in the timeline that just contains a timeline array and no other parameters.

I think the reason that your code wasn't working before is because the ReadAllResponses object needs to be created after the two trials that it contains have been defined. So it should work if you just reorder the definitions of the trials and the ReadAllResponses object:

var ReadOneResponse = {

  type: 'audio-button-response',

  stimulus: RecallSignal,

  button_html: ListOfButtons,

  choices: function() {return [...Array(ListOfButtons.length).keys()]},

  trial_duration: 5000,

  on_finish: function() {console.log(jsPsych.data.get().last(1).values()[0].button_pressed)}

}

var BlinkThatResponse = {

  type: 'html-button-response',

  stimulus: '',

  button_html: function() {return ListOfButtons.concat(ListOfItems[jsPsych.data.get().last(1).values()[0].button_pressed])},

  trial_duration: 1000,

  choices: function() {return [...Array(ListOfButtons.length+1).keys()]}

}

var ReadAllResponses = {

  timeline: [ ReadOneResponse, BlinkThatResponse ]

}

jsPsych.init({

  timeline: [ ReadAllResponses ]

})

So I'm going to unmark my other response and mark this one as the answer. But please let me know if you get a chance to try this and it doesn't solve the problem!
